### DISCUSSION TITLE
why machine id is different after reboot

### DISCUSSION THREAD

**Discussion by Tianjie-Wu** [dashingsoft/pyarmor#1378] [Q&A] [answered]
Hello, I've observed that the machine ID to which the pyarmor group license was previously tied has changed after a reboot. And this will make the license on this machine invalid.
Then I conducted a small experiment myself. Repeatedly reboot the machine and record the machine ID after each reboot.
On this computer, six unique IDs will be generated. At each reboot, pyarmor will choose one at random.

After 19 reboots, the ID I registered appears.
I believe there is at least one flaw in the machine ID generation process.

PS: I'm aware that unlimited docker can operate under a group license, but all docker must run on a host machine; regrettably, in this case, the host machine is unable to maintain a fixed ID.
And I already tried `pyarmor-7 hdinfo` , everytime is the same.

为了避免误解，我用中文再说一遍：
我发现注册好集团版许可证的机器，一旦重启，machine ID 会变，导致注册的许可失效。
我测试了一下，多次重启机器，并记录ID
我发现在我这台机器上，会生成6个不同的ID，每次重启后，运行pyarmor，他会从中选一个作为ID，并再下一次重启前保持不变。

我重启了19次，才获得到和原来一样的ID。
所以我觉得生成ID的地方可能会有一些问题。

PS：我知道集团版可以运行不限制数量的docker,但是docker依赖于已经注册的pyarmor，但是不幸的是，这台总是变来变去ID的机器就是运行docker的主机。
我尝试过pyarmor-7 hdinfo，但是每次返回的值是一样的。

attached is the script to get machine id 
我用下面的脚本获取machine id

```bash
cd /tmp
pyarmor reg -g 88
cat .pyarmor/group/pyarmor-group-device.88
```

attached is the ID that appeared on that machine:
我把出现过的ID列在下面
每次出现的很随机，我也没办法选择最常出现的ID注册

```
b8bf588e69653469e73d********c300f
bdb6da353da1e05edb06f********848e  (* 2 times )
b338f821ef57fae11901b********4fab  (* 3 times)
b8820e8403dd55dbb0cdb********c08c (* 5 times)
bf13e4b16d172ee1cc936********8a0e (* 3 times)
b1a34320909d370534489********ec06 (* 5 times)
```

**Comment by jondy** [comment]
你使用的 Pyarmor 的版本呢？

**Comment by Tianjie-Wu** [comment]
I am using 8.2.9

```
Pyarmor 8.2.9 (group), 00****, *******

License Type    : pyarmor-group
License No.     : pyarmor-vax-00****
License To      : ***
License Product : *******

BCC Mode        : Yes
RFT Mode        : Yes

Notes
* Offline obfuscation
```

**Comment by jondy** [comment]
操作系统是什么？

**Reply by Tianjie-Wu** [reply]
操作系统是：
```
Red Hat Enterprise Linux release 8.8 (Ootpa)
Linux program-terminal 4.18.0-477.13.1.el8_8.x86_64 #1 SMP Thu May 18 10:27:05 EDT 2023 x86_64 x86_64 x86_64 GNU/Linux
```

**Comment by jondy** [comment]
这台机器是物理机，虚拟机，还是云服务器？如果是虚拟机的话，是什么虚拟机（VirtualBox？Qemu?...)

**Reply by Tianjie-Wu** [reply]
这台机器是我们单位资源池中的机器，他们使用的是ESXI（基于vmware?）, 应该算是虚拟机，给的配置是8核32G。
我检查了一下，在/dev/by-id的磁盘ID是一致的，应该也是虚拟磁盘，容量为1T，hostname每次也不会变化，IP地址是和MAC绑定的也没变过。

**Comment by jondy** [comment]
有没有任意一个网络接口 MAC address 的地址不是 12 位的？例如 只有 10位 `01:23:45:67:89`

**Reply by Tianjie-Wu** [reply]
看上去没有：
```
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host 
       valid_lft forever preferred_lft forever
2: ens192: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc mq state UP group default qlen 1000
    link/ether 00:0c:29:8d:**:c5 brd ff:ff:ff:ff:ff:ff
    altname enp11s0
    inet 192.168.88.8/24 brd 192.168.88.255 scope global dynamic noprefixroute ens192
       valid_lft 81283sec preferred_lft 81283sec
    inet6 fe80::7d15:e344:9617:e968/64 scope link noprefixroute 
       valid_lft forever preferred_lft forever
3: ens224: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc mq state UP group default qlen 1000
    link/ether 00:0c:29:8d:**:cf brd ff:ff:ff:ff:ff:ff
    altname enp19s0
    inet 192.168.9.8/24 brd 192.168.9.255 scope global dynamic noprefixroute ens224
       valid_lft 81283sec preferred_lft 81283sec
    inet6 fe80::d2c2:6622:5717:e500/64 scope link noprefixroute 
       valid_lft forever preferred_lft forever
4: ztukuq3723: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 2800 qdisc fq_codel state UNKNOWN group default qlen 1000
    link/ether 6a:e6:4e:a0:**:c0 brd ff:ff:ff:ff:ff:ff
    inet 192.168.192.8/24 brd 192.168.192.255 scope global ztukuq3723
       valid_lft forever preferred_lft forever
    inet6 fe80::68e6:4eff:fea0:12c0/64 scope link 
       valid_lft forever preferred_lft forever
5: virbr0: <NO-CARRIER,BROADCAST,MULTICAST,UP> mtu 1500 qdisc noqueue state DOWN group default qlen 1000
    link/ether 52:54:00:f0:**:48 brd ff:ff:ff:ff:ff:ff
    inet 192.168.122.1/24 brd 192.168.122.255 scope global virbr0
       valid_lft forever preferred_lft forever
```

**Comment by jondy** [comment]
所有接口的 mac addr 重启后都没有变化吗？尤其这个 `virbr0`

**Reply by Tianjie-Wu** [reply]
我重启了五次，确认mac没有变化
virbr0是docker（redhat里叫podman）的接口，似乎不会变

**Comment by SethXie** [comment]
我也是遇到类似的问题，因为注意到几次machine id会变化，所以提了这个https://github.com/dashingsoft/pyarmor/discussions/1373

我虚拟机用的是proxmox
Mac、IP、hostname都没改的情况下，machine id也变了。
用./hdinfo工具看是这样的：
```
Hardware informations got by PyArmor:

Failed to get harddisk information.

Default Mac address: "06:64:4b:78:8f:**"

Multiple Mac addresses: "<06:64:4b:78:8f:**,02:42:c0:62:4e:**,02:42:a1:48:42:**,02:42:2c:24:d7:**,96:48:84:80:e4:**>"

Ip address: "192.168.100.**"

Domain name: "(none)"

Change logs

        v6.2.0(r21): Remove trailing dot from harddisk serial number
        v6.4.2(r34): Support binding multiple mac addresses
        v6.5.3(r37): Support binding named harddisk
        v6.7.5(r45): Support mmc/sd card in Linux

```
最开始注册机器生成machine id的时候pyarmor是8.2.0；
目前pyarmor版本是8.2.9，但是我尝试过把它卸载重装 8.2.0，跟跟当时的machine id也是不一样。
默认Mac地址肯定没变，但其他的地址是否变化我没有注意。假如变了的话，这个也会影响machine id么？

**Reply by Tianjie-Wu** [reply]
我测试了IP地址变化好像没啥问题
但是hostname不能变
不知道是不是个例。

然后目前使用这个脚本，让机器重启直到获得和注册信息一样的ID才停。
脚本有风险，仅供参考

```bash
source ~/.bashrc
conda activate pyarmor
cd /tmp
pyarmor reg -g 88
date
my_id_now=`cat .pyarmor/group/pyarmor-group-device.88|awk '{print $2}'`
if [ ${my_id_now} = "b8bf588e69653469e73d********c300f" ]
then
        echo "same id"
else
        echo "different id, reboot now!"
        sleep 30s
        sudo reboot
fi
```

用crontab让他每次重启后运行：

```
@reboot sleep 30s && sh /home/tianjie/QuickLogin/pyarmor_id.sh >> /home/tianjie/.pyarmor_reboot_log
```

**Reply by SethXie** [reply]
我hostname一直都没变过。这个脚本倒是一种方法，但是万一一直没有得到注册过的id，机器岂不是一直在重启。还是想了解一下具体都有哪些因素影响machine id，毕竟group license有100台设备的限制。

**Reply by Tianjie-Wu** [reply]
你误会了，这个脚本只能是确定我这种情况的才能使用，而且平时要把reboot那一行注释掉才行。只有确定要用了才去掉reboot的注释。我想肯定有更好的解决方法，只不过我这是临时解决问题而已。至于你说到底有哪些方面会导致ID变化，我只知道mac和hostname，主要是每次重启变ID有点让我没想到...

**Comment by jondy** [comment]
目前来说所有的 MAC address 都会有影响的，machine id 就是 8.2.9 的算法改变了一次，但是以前的 machine id 在新版本里面是兼容的。

不同操作系统的 machine id 算法不一样，但是基本都需要 hostname, 所有接口的 MAC address，以及 硬盘信息。

**Reply by Tianjie-Wu** [reply]
就是比较奇怪的是，这些一般esxi(或者说vmware)的虚拟机一般是生成就不变了，我记录了每次重启的MAC，hostname , /dev/disk/by-id /dev/disk/by-uuid ，似乎都没有变过，我实在是不太明白vmware到底每次重启干了啥。我们也给vmware发了邮件了，正在等他们的回复。

**Comment by SethXie** [comment]
> 目前来说所有的 MAC address 都会有影响的

@jondy 所有mac地址，包括bridge和veth吧？是否意味着：我在宿主机上新启动或者删掉一些docker container，也会导致host机器上的mac地址有所变化，从而导致machine id变化？

**Reply by jondy** [reply, marked as answer]
如果网络配置增加/减少/修改 Mac Address，machine id 会发生变化的。
8.3.0 里面可能会考虑 集团版许可证使用单独的算法，排除 bridge 等的影响。

**Reply by Tianjie-Wu** [reply]
这对我是一个很好的消息，谢谢！

**Comment by jondy** [comment]
8.3.dev1 已经发布，集团版许可证使用新的 machine id 算法 `get_hd_info(16)`，忽略虚拟网卡的影响。

**Reply by Tianjie-Wu** [reply]
我刚刚测试了一下，的确现在旧版的ID变化的时候，新ID不会变化了
